text/0000-generic-int.md

```diff
@@ -0,0 +1,552 @@
+primitives. The standard library itself is a great example; almost the entirety
+of `core::num` uses some gnarly macros to define all sorts of traits for all the
+integer types. One example is `Shl` and `Shr`, which are defined for not just
+every integer type, but every *combination* of integer types. We could easily do
+this with const generics instead:
+
+```rust
+impl<const N: usize, const M: usize> Shl<uint<M>> for uint<N> {
+    type Output = uint<N>;
+    #[inline]
+    fn shl(self, rhs: uint<M>) -> unit<N> {
+        // implementation
+    }
+}
+```
+
+This will probably also decrease compiler time as well, as the compiler simply
+needs to understand one generic impl instead of 144 macro-generated impls.
+Because `Shl` and `Shr` can also handle references, that multiplies our
+generated impl count by four per trait, meaning that our macros generate 576
+impls total!
+
+## Enum optimisations
+
+Using a smaller number of bits for a value also has the benefit of niche enum
+optimisations. For example, `uint<7>` represents a single ASCII character, and
+`Option<uint<7>>` can be stored in a single byte. Additionally,
+`Result<uint<7>, E>` also takes one byte if `E` is zero-sized.
+
+## Documentation
+
+Having generic impls would drastically reduce the noise in the "implementations"
+section of rustdoc. For example, the number of implementations for `Add` for
+integer types really drowns out the fact that it's also implemented for strings
+and `std::time` types, which is useful to know too.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Up until now, we've only seen *specific* integer types, like `u8` or `i32`. What
+if we want to write a trait that works with all of them?
+
+Let's go with a simple example. Let's say that we want to code our own `Default`
+trait which returns one instead of zero. We'll define it like this:
+
+```rust
+pub trait MyDefault: Sized {
+    fn my_default() -> Self;
+}
+```
+
+Right now, with what we know, we have to copy-paste our impl for every integer
+type. It would look something like:
+
+```rust
+impl MyDefault for u8 {
+    fn my_default() -> u8 { 1 }
+}
+```
+
+Except, we'd have to replicate this for every single integer type. If we're
+clever, we could use a macro:
+
+```rust
+macro_rules! impl_my_default {
+    ($($int: ident),*) => {$(
+        impl MyDefault for $int {
+            fn my_default() -> $int { 1 }
+        }
+    )*}
+}
+impl_my_default!(i8, u8, i16, u16, i32, u32, i64, u64, i128, u128, isize, usize);
+```
+
+But really, that's just making the compiler do the copy-pasting for us. Instead,
+we're going to use the special types `uint<N>` and `int<N>` to generalise the
+code for us. For this, we're going to use const generics, instead of type
+generics. The end result looks like:
+
+```rust
+impl<const N: usize> MyDefault for uint<N> {
+    fn my_default() -> uint<N> { 1 }
+}
+impl<const N: usize> MyDefault for int<N> {
+    fn my_default() -> int<N> { 1 }
```

Comment:
Since 1 isn't in range for `int<1>`, is this a monomorphization time error?  It seems like a lot of seeminly reasonable impls would only work for a minimum size.

Reply:
These constraints allow `N == 0` as well.

Reply:
I was under the impression that overflowing literals was merely a lint that could be allowed. So, the only change necessary here is `#[allow(overflowing_literals)]`.

Reply:
Doesn't `#[allow(overflowing_literals)]` only hide (what might be) a genuine problem rather than fixing the issue?